kernel/src/sched.rs

```diff
@@ -524,6 +520,55 @@ impl Kernel {
         }
     }
 
+    /// Main loop of the OS.
+    ///
+    /// Most of the behavior of this loop is controlled by the `Scheduler`
+    /// implementation in use.
+    pub fn kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>, const NUM_PROCS: usize>(
+        &self,
+        platform: &P,
+        chip: &C,
+        ipc: Option<&ipc::IPC<NUM_PROCS>>,
+        scheduler: &SC,
+        capability: &dyn capabilities::MainLoopCapability,
+    ) -> ! {
+        chip.watchdog().setup();
+        loop {
+            self.loop_op(platform, chip, ipc, scheduler, capability, false);
+        }
+    }
+
+    /// Main test loop of the OS.
+    ///
+    /// This is similar to `kernel_loop()` but also calls tests passed in via
+    /// the `tests` argumet. This should only be used for running tests.
+    pub fn test_kernel_loop<P: Platform, C: Chip, SC: Scheduler<C>, const NUM_PROCS: usize>(
+        &self,
+        platform: &P,
+        chip: &C,
+        ipc: Option<&ipc::IPC<NUM_PROCS>>,
+        scheduler: &SC,
+        capability: &dyn capabilities::MainLoopCapability,
+        tests: &[&dyn Fn()],
+    ) {
+        chip.watchdog().setup();
+        // Run all of the tests
+        for test in tests {
+            for _ in 0..20 {
+                self.loop_op(platform, chip, ipc, scheduler, capability, true);
+            }
+            test();
+            for _ in 0..100 {
+                self.loop_op(platform, chip, ipc, scheduler, capability, true);
+            }
+        }
```

Comment:
From a design point, I don't really understand why this loop is here? In particular, the loops of 20 and 100 seem somewhat odd and arbitrary. It feels like this loop should really be the responsibility of the test harness runner.

Reply:
The numbers are arbitrary. As long as the loop runs enough times that the test can complete and all debug output is printed (but not an unlimited number of times) it doesn't really matter. We could move this to the test harness (in the board), but this should be the same across boards and this way we aren't exposing the internal kernel loop logic to boads

Reply:
It does seem to be the case that different tests might require different numbers of iterations to complete, and once some longer running tests are included it will make sense to not make every test wait the length of time required by the longest test. Though maybe none of our tests will ever be long enough for that to matter